pkg/maps/utime/utime.go

```diff
@@ -0,0 +1,244 @@
+	// nanoseconds by 2^9 yields roughly half microsecond accuracy but avoids expensive 64-bit
+	// divisions in the datapath.  With this shift the range of an u64 is ~300000 years
+	// instead of ~600 years if left at nanoseconds.
+	// With this shift full seconds can be multiplied by 1e9>>9 to get utime units.
+	// Must be kept in sync with `UTIME_SHIFT` in datapath (bpf/lib/utime.h), any changes to
+	// this will have an upgrade impact.
+	utimeShift = 9
+	// 10^9 has 9 trailing zeroes also in binary, so they can be shifted off without any loss
+	// of accuracy.
+	secsToUtimeMultiplier = 1000_000_000 >> utimeShift // integer value (1953125)
+	// utime numerical limits in seconds
+	minSeconds = 0
+	maxSeconds = 1 << (64 + utimeShift) / 1_000_000_000 // 2^(64+9)/10^9
+)
+
+// UTimeKey is the index to the Unix time offset. Only value 0 is supported for now.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type UTimeKey struct {
+	key uint32 `align:"ctx"`
+}
+
+// Unix epoch time value on 2^9/10^9 second accuracy. This accuracy
+// is chosen so that the timing is reasonably accurate for expiry times, but does not require 64-bit
+// division of a monotonic clock value in the datapath, as it is a rather slow operation.
+type UTime uint64
+
+func toUTime(secs int64, nanos int) UTime {
+	return UTime(secs)*secsToUtimeMultiplier + UTime(nanos)>>utimeShift
+}
+
+func TimeToUTime(t time.Time) UTime {
+	return toUTime(t.Unix(), t.Nanosecond())
+}
+
+func (t UTime) Time() time.Time {
+	secs := t / secsToUtimeMultiplier
+	usecs := t % secsToUtimeMultiplier
+	return time.Unix(int64(secs), int64(usecs<<utimeShift))
+}
+
+// UtimeValue is the offset to be added to the monotonic time to get the Unix epoch time.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type UTimeValue struct {
+	uTime UTime
+}
+
+// String pretty print the UTimeKey
+func (k UTimeKey) String() string {
+	return fmt.Sprintf("%d", k.key)
+}
+
+// String pretty print the UTimeValue.
+func (v UTimeValue) String() string {
+	return fmt.Sprintf("%s", v.uTime.Time().String())
+}
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *UTimeValue) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k *UTimeKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// NewValue returns a new empty instance of the structure represeting the BPF
+// map value
+func (k UTimeKey) NewValue() bpf.MapValue { return &UTimeValue{} }
+
+func newUTimeKey(key uint32) *UTimeKey {
+	return &UTimeKey{
+		key: key,
+	}
+}
+
+const (
+	// MapName name of map used to pin map for datapath
+	MapName = "cilium_time_offset"
+
+	// MaxEntries represents the maximum number of time offsets
+	MaxEntries = 1
+)
+
+type UTimeMap struct {
+	*bpf.Map
+}
+
+var (
+	once     sync.Once
+	uTimeMap UTimeMap
```

Comment:
Let's please make this into a cell from the start and not add more global variables.

Reply:
(Please capture the output of any Slack discussion or missing docs in https://docs.cilium.io/en/latest/contributing/development/hive/#id1 , I would like to make sure this information is publicly available and doesn't require any contributor to schedule a synchronous session with a committer in order to make progress)

Reply:
@mhofstetter volunteered to transition this to a Hive Cell in an "immediate" follow-up PR. He did the same for the auth map so that would fit me perfectly :-)

Reply:
awesome thanks @mhofstetter

Reply:
Unfortunately none of the `pkg/maps` are defined as Cells yet, so there is no example to go by.

Reply:
branch with a initial proposal is ready - will open a PR as soon as this PR is merged